drivers/gpio/gpio_pca95xx.c

```diff
@@ -0,0 +1,671 @@
+				"(%d)", i2c_addr, reg, ret);
+	}
+	return ret;
+}
+
+/**
+ * @brief Setup the pin direction (input or output)
+ *
+ * @param dev Device struct of the PCA95XX
+ * @param access_op Access operation (pin or port)
+ * @param pin The pin number
+ * @param flags Flags of pin or port
+ *
+ * @return 0 if successful, failed otherwise
+ */
+static int setup_pin_dir(struct device *dev, int access_op,
+			  u32_t pin, int flags)
+{
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	union gpio_pca95xx_port_data *reg_dir = &drv_data->reg_cache.dir;
+	union gpio_pca95xx_port_data *reg_out = &drv_data->reg_cache.output;
+	int ret;
+
+	/* For each pin, 0 == output, 1 == input */
+	switch (access_op) {
+	case GPIO_ACCESS_BY_PIN:
+		if ((flags & GPIO_OUTPUT) != 0U) {
+			if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0U) {
+				reg_out->all |= BIT(pin);
+			} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0U) {
+				reg_out->all &= ~BIT(pin);
+			}
+			reg_dir->all &= ~BIT(pin);
+		} else {
+			reg_dir->all |= BIT(pin);
+		}
+
+		break;
+	case GPIO_ACCESS_BY_PORT:
+		if ((flags & GPIO_OUTPUT) != 0U) {
+			if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0U) {
+				reg_out->all = 0xFFFF;
+			} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0U) {
+				reg_out->all = 0x0;
+			}
+			reg_dir->all = 0x0;
+		} else {
+			reg_dir->all = 0xFFFF;
+		}
+
+		break;
+	default:
+		ret = -ENOTSUP;
+		goto done;
+	}
+
+	ret = write_port_regs(dev, REG_OUTPUT_PORT0, reg_out);
+	if (ret != 0) {
+		goto done;
+	}
+
+	ret = write_port_regs(dev, REG_CONF_PORT0, reg_dir);
+
+done:
+	return ret;
+}
+
+/**
+ * @brief Setup the pin pull up/pull down status
+ *
+ * @param dev Device struct of the PCA95XX
+ * @param access_op Access operation (pin or port)
+ * @param pin The pin number
+ * @param flags Flags of pin or port
+ *
+ * @return 0 if successful, failed otherwise
+ */
+static int setup_pin_pullupdown(struct device *dev, int access_op,
+				 u32_t pin, int flags)
+{
+	const struct gpio_pca95xx_config * const config =
+		dev->config->config_info;
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	union gpio_pca95xx_port_data *port;
+	u16_t bit_mask;
+	u16_t new_value = 0U;
+	int ret;
+
+	if ((config->capabilities & PCA_HAS_PUD) == 0) {
+		/* Chip does not support pull up/pull down */
+		if ((flags & (GPIO_PULL_UP | GPIO_PULL_DOWN)) != 0U) {
+			ret = -ENOTSUP;
+			goto done;
+		}
+
+		/* If both GPIO_PULL_UP and GPIO_PULL_DOWN are not set,
+		 * we should disable them in hardware. But need to skip
+		 * if the chip does not support pull up/pull down.
+		 */
+		ret = 0;
+		goto done;
+	}
+
+	/* If disabling pull up/down, there is no need to set the selection
+	 * register. Just go straight to disabling.
+	 */
+	if ((flags & (GPIO_PULL_UP | GPIO_PULL_DOWN)) == 0U) {
+		goto en_dis;
+	}
+
+	/* Setup pin pull up or pull down */
+	port = &drv_data->reg_cache.pud_sel;
+	switch (access_op) {
```

Comment:
As with previous function: validate against and implement `GPIO_ACCESS_BY_PIN` only.

Also applies to switch statement later in this function.

Reply:
(as above about `gpio_config()` is still a valid public API.)